base/inference.jl

```diff
@@ -2120,13 +2121,139 @@ function inlineable(f, e::Expr, atypes, sv, enclosing_ast)
     if isa(f,IntrinsicFunction)
         return NF
     end
+    if is(f, invoke)
+        return inlineable_invoke(f, e, atypes, sv, enclosing_ast, argexprs)
+    end
+
+    return inlineable_gf(f, e, atypes, sv, enclosing_ast, argexprs)
+end
+
+function get_invoke_types{T<:Top}(ts::Type{Type{T}})
+    return T
+end
+
+function get_invoke_types(ts::Tuple)
+    return tuple(Any[get_invoke_types(t) for t in ts]...)::Type
+end
+
+function inlineable_invoke(f::Function, e::Expr, atypes, sv,
+                           enclosing_ast, argexprs)
+    f = isconstantfunc(argexprs[1], sv)
+    if is(f, false)
+        return NF
+    end
+    f = _ieval(f)
+    local invoke_types
+    try
+        invoke_types = get_invoke_types(atypes[2])
+    catch err
+        return NF
+    end
 
+    argexprs = argexprs[3:end]
+    atypes = atypes[3:end]
+    if length(atypes) != length(invoke_types)
+        return NF
+    end
+    if isleaftype(invoke_types) && invoke_types == atypes
+        new_e = Expr(:call, f, argexprs...)
+        new_e.typ = e.typ
+        res = inlineable_gf(f, new_e, invoke_types, sv,
+                            enclosing_ast, argexprs)
+        return is(res, NF) ? (new_e, []) : res
+    end
+
+    local meth
+    try
+        meth = ccall(:jl_gf_invoke_lookup, Any, (Any, Any), f, invoke_types)
+    catch
+        return NF
+    end
+    if is(meth, nothing)
+        return NF
+    end
+
+    stmts = []
+    check_stmts = []
+    atypes_l = Type[atypes...]
+    err_label = genlabel(sv)
+    after_err_label = genlabel(sv)
+    for i in 1:length(atypes_l)
```

Comment:
I think this is what I'm trying to do here. I've seen that part but I didn't really understand why. Isn't the arguments evaluated in the order they appears in the code?

Also the code generated seems to be currect here and that's why I didn't bother too much before sending the PR
```julia
julia> @noinline function get_next()
           global counter
           counter = counter + 1
           return counter
       end
get_next (generic function with 1 method)

julia> f(a, b) = (a, b)
f (generic function with 1 method)

julia> g() = invoke(f, (Integer, Integer), get_next(), get_next())
g (generic function with 1 method)

julia> @code_typed g()
1-element Array{Any,1}:
 :($(Expr(:lambda, Any[], Any[Any[:_var0,:_var1],Any[Any[:_var0,Any,18],Any[:_var1,Any,18]],Any[]], :(begin  # none, line 1:
        _var0 = get_next()::Any
        _var1 = get_next()::Any
        unless (isa)(_var0,Integer)::Bool goto 1
        unless (isa)(_var1,Integer)::Bool goto 1
        goto 2
        1: 
        (error)("invoke: argument type error")::Union()
        2: 
        return (top(tuple))(_var0::Integer,_var1::Integer)::(Integer,Integer)
    end::(Integer,Integer)))))

julia> g()
ERROR: counter not defined
 in get_next at ./no file:3
 in g at ./none:1

julia> counter = 0
0

julia> g()
(1,2)
```